fairchive-webapp/src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/impl/PBIOauth2AP.java

```
@@ -0,0 +1,152 @@
+                    OAuth2TokenData.from(token),
+                    new AuthenticatedUserDisplayInfo(
+                            json.get("given_name").getAsString(),
+                            json.get("family_name").getAsString(),
+                            json.get("email").getAsString(), "", ""),
+                    singletonList(json.get("email").getAsString()));
+
+        } catch (final ExecutionException | InterruptedException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public String createState(final Optional<String> redirectPage) {
+        try {
+            final String verifier = randomAlphanumeric(50);
+            final String challenge = PKCECodeChallengeMethod.S256
+                    .transform2CodeChallenge(verifier);
+            final String encrytpted = StringUtil.encrypt(verifier + "~" + challenge,
+                    getClientSecret());
+            return getId() + "~" + encrytpted;
+        } catch (final NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String getChallengeFrom(final String state) {
+        final String enctypted = state.substring(getId().length() + 1);
+        return StringUtil.decrypt(enctypted, getClientSecret()).split("~")[1];
+    }
+
+    private String getVerifierFrom(final String state) {
+        final String enctypted = state.substring(getId().length() + 1);
+        return StringUtil.decrypt(enctypted, getClientSecret()).split("~")[0];
+    }
+
+    @Override
+    public boolean isDisplayIdentifier() {
+        return false;
+    }
+
+    @Override
+    protected ParsedUserResponse parseUserResponse(String responseBody) {
+        throw new RuntimeException("Not implemented.");
+    }
+
+    private static JsonObject extractPayloadFrom(final String jwt) {
+        final int firstDotIndex = jwt.indexOf('.');
+        final int secondDotIndex = jwt.indexOf('.', firstDotIndex + 1);
```

According to documentation JWT token has 3 parts https://www.jwt.io/introduction#what-is-json-web-token-structure
Maybe we should throw expetion when second dot not exists because it's invalid jwt token?

---

check added